proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java

```diff
@@ -0,0 +1,157 @@
+    private final int broadcastPort;
+    private final int receiverBufferSize = 512;
+    private final DatagramSocket datagramSocket;
+    private final UdpBeaconDecoder udpBeaconDecoder;
+    private int maxAttemptsNumber;
+    private int attemptsPerformed = 0;
+    private String pushAddress;
+    private String subscribeAddress;
+
+    public UdpBeaconFinder(final DatagramSocket datagramSocket,
+                           final int broadcastPort,
+                           final UdpBeaconDecoder udpBeaconDecoder) {
+        this.datagramSocket = datagramSocket;
+        this.broadcastPort = broadcastPort;
+        this.udpBeaconDecoder = udpBeaconDecoder;
+        maxAttemptsNumber = 1;
+    }
+
+    /**
+     * Change number of attempts to find the UDP beacon server
+     * Default value is set to 1
+     * @param maxAttemptsNumber
+     */
+    public void setMaxAttemptsNumber(final int maxAttemptsNumber) {
+        if (0 > maxAttemptsNumber) {
+            logback.warn("Udp broadcasting attempts number cannot be less than 0");
+        } else {
+            this.maxAttemptsNumber = maxAttemptsNumber;
+        }
+    }
+
+    public int getNumberOfPerformedAttempts() {
+        return attemptsPerformed;
+    }
+
+    /**
+     * @return true if we did not get any meaningful response from server
+     * and have not reached max allowed attempts number
+     */
+    private boolean shouldTryToFindBeacon() {
+        if (null == udpBeaconDecoder) {
+            return attemptsPerformed < maxAttemptsNumber;
+        }
+        return (!udpBeaconDecoder.hasReceivedCorrectData() &&
+                attemptsPerformed < maxAttemptsNumber);
+    }
+
+    /**
+     * Find the server using UDP broadcast and fill data fields
+     */
+    public void broadcastAndGetServerAddress() {
+        try {
+            datagramSocket.setBroadcast(true);
+            while (shouldTryToFindBeacon()) {
+                logback.info("Trying to find UDP beacon, attempt [" + new Integer(attemptsPerformed+1) + "]");
+                // Broadcast the message over all the network interfaces
+                final Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
+                while (interfaces.hasMoreElements()) {
+                    final NetworkInterface networkInterface = interfaces.nextElement();
+
+                    if (networkInterface.isLoopback() || !networkInterface.isUp()) {
+                        continue; // Do not broadcast to the loopback interface
+                    }
+
+                    for (final InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()) {
+                        final InetAddress broadcastAddress = interfaceAddress.getBroadcast();
+                        if (null != broadcastAddress) {
+                            logback.info("Trying to send broadcast package on interface: " + networkInterface.getDisplayName());
```

Comment:
I think there was a comment earlier about how far this goes but It really is indented a lot.